File: web/lib/db/publish-attempts.ts
import { createClient } from "@/lib/supabase/server";
import type { Database } from "@/lib/db/types";
import type { SupabaseClient } from "@supabase/supabase-js";

type PublishAttemptRow =
  Database["public"]["Tables"]["publish_attempts"]["Row"];
type PublishAttemptInsert =
  Database["public"]["Tables"]["publish_attempts"]["Insert"];

type AnyClient = SupabaseClient<Database>;

export async function createPublishAttempt(
  values: PublishAttemptInsert,
  client?: AnyClient
): Promise<PublishAttemptRow> {
  const supabase = client ?? (await createClient());
  const { data, error } = await supabase
    .from("publish_attempts")
    .insert(values)
    .select()
    .single();
  if (error) throw error;
  return data;
}

export async function updatePublishAttempt(
  id: string,
  patch: Partial<PublishAttemptRow>,
  client?: AnyClient
): Promise<void> {
  const supabase = client ?? (await createClient());
  const { error } = await supabase
    .from("publish_attempts")
    .update(patch)
    .eq("id", id);
  if (error) throw error;
}

export async function getLatestAttempt(
  postVariantId: string,
  client?: AnyClient
): Promise<PublishAttemptRow | null> {
  const supabase = client ?? (await createClient());
  const { data, error } = await supabase
    .from("publish_attempts")
    .select("*")
    .eq("post_variant_id", postVariantId)
    .order("attempt_number", { ascending: false })
    .limit(1)
    .maybeSingle();
  if (error) return null;
  return data;
}

// Idempotency guard: has this variant already been successfully published?
export async function hasSuccessfulAttempt(
  idempotencyKey: string,
  client?: AnyClient
): Promise<boolean> {
  const supabase = client ?? (await createClient());
  const { count, error } = await supabase
    .from("publish_attempts")
    .select("*", { count: "exact", head: true })
    .eq("idempotency_key", idempotencyKey)
    .eq("status", "success");
  if (error) return false;
  return (count ?? 0) > 0;
}

File: worker/db/publish_attempts.py
from typing import Any

from supabase import AsyncClient


async def create_publish_attempt(
    client: AsyncClient, values: dict[str, Any]
) -> dict[str, Any]:
    res = await client.table("publish_attempts").insert(values).execute()
    return res.data[0]


async def update_publish_attempt(
    client: AsyncClient, attempt_id: str, patch: dict[str, Any]
) -> None:
    await client.table("publish_attempts").update(patch).eq(
        "id", attempt_id
    ).execute()


async def get_latest_attempt(
    client: AsyncClient, post_variant_id: str
) -> dict[str, Any] | None:
    res = (
        await client.table("publish_attempts")
        .select("*")
        .eq("post_variant_id", post_variant_id)
        .order("attempt_number", desc=True)
        .limit(1)
        .maybe_single()
        .execute()
    )
    return res.data


async def has_successful_attempt(
    client: AsyncClient, idempotency_key: str
) -> bool:
    res = (
        await client.table("publish_attempts")
        .select("id", count="exact", head=True)
        .eq("idempotency_key", idempotency_key)
        .eq("status", "success")
        .execute()
    )
    return (res.count or 0) > 0
